enabler/src/de/schildbach/pte/HslProvider.java

```diff
@@ -0,0 +1,833 @@
+/*
+ * Copyright 2010-2015 the original author or authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package de.schildbach.pte;
+
+import java.net.URLEncoder;
+import java.io.IOException;
+import java.io.InputStream;
+import java.text.ParsePosition;
+import java.text.SimpleDateFormat;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Date;
+import java.util.EnumSet;
+import java.util.GregorianCalendar;
+import java.util.HashMap;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.HashSet;
+
+import com.google.common.base.Joiner;
+
+import de.schildbach.pte.dto.Departure;
+import de.schildbach.pte.dto.Line;
+import de.schildbach.pte.dto.Location;
+import de.schildbach.pte.dto.LocationType;
+import de.schildbach.pte.dto.NearbyLocationsResult;
+import de.schildbach.pte.dto.Point;
+import de.schildbach.pte.dto.Product;
+import de.schildbach.pte.dto.QueryDeparturesResult;
+import de.schildbach.pte.dto.QueryTripsContext;
+import de.schildbach.pte.dto.QueryTripsResult;
+import de.schildbach.pte.dto.ResultHeader;
+import de.schildbach.pte.dto.StationDepartures;
+import de.schildbach.pte.dto.Stop;
+import de.schildbach.pte.dto.SuggestLocationsResult;
+import de.schildbach.pte.dto.SuggestedLocation;
+import de.schildbach.pte.dto.Trip;
+import de.schildbach.pte.dto.Style;
+
+import de.schildbach.pte.exception.ParserException;
+
+import de.schildbach.pte.util.HttpClient;
+import de.schildbach.pte.util.XmlPullUtil;
+
+import org.xmlpull.v1.XmlPullParser;
+import org.xmlpull.v1.XmlPullParserException;
+import org.xmlpull.v1.XmlPullParserFactory;
+
+/**
+ * @author Mats Sjöberg <mats@sjoberg.fi>
+ */
+public class HslProvider extends AbstractNetworkProvider 
+{
+	private static final int COORD_MUL = 1000000;
+	private static final String SERVER_PRODUCT = "hsl";
+	private static final String SERVER_VERSION = "1_2_0";
+	private static final int EARLIER_TRIPS_MINUTE_OFFSET = 5;
+	private static final int EARLIER_TRIPS_MINIMUM = 3;
+
+	private String apiBase;
+
+	private final XmlPullParserFactory parserFactory;
+
+	public HslProvider(String user, String pass)
+	{
+		super(NetworkId.HSL);
+		apiBase = String.format("http://api.reittiopas.fi/hsl/%s/?user=%s&pass=%s",
```

Comment:
Why are username and password part of the API base? My understanding is that they are constant (http://api.reittiopas.fi/hsl/ in this case) and the parameters are appended dynamically. Username and password could be appended in the apiUri() method.